fw_upgrade/README.md

```diff
@@ -0,0 +1,123 @@
+# Firmware Upgrade
+
+Upgrading the router's firmware usually works via NetCloud, for newer firmware releases (`> 2018`) this is the only way. For older releases upgrading via `scp` or the web interface still works.
+
+Older firmware images are still available on https://cradlepoint.com. They are signed with a RSA private key and encrypted with AES256-CBC. However, the signature verification can be easily bypassed and the AES key is hard-coded in the router's Python middleware.
+
+## Implementation
+
+The firmware upgrade is mainly implemented in the `/service_manager/services/utils/upgrade_write.pyc` file. This is what's going on:
+
+1. `upgrade_write()` --> decrypts the upgrade file, writes a block, gets the `uimage` and `uboot` signatures from the firmware file *or* calls `upgrade_write_init()`
+2. `upgrade_write_init()` --> header and CRC checks, writes a new SHA512 hash to flash memory
+3. `_write()` --> writes data to memory flash
+4. `upgrade_validate()` --> validates the file and checks the signature by calling `upgrade_validate_signature()`
+
+If the version string in the firmware header matches a certain value, the signature verification can be bypassed:
+
+```python
+if upgrade_int >= 458752:
+    self.force_signature_validation = True
+```
+
+## Get the AES Decryption Key
+
+Firmware update files which are available on the Cradlepoint website are encrypted with AES. However, at least for the `IBR600` a global hard-coded AES key is used.
+
+In a nutshell, the key is generated with OpenSSL (in `/usr/lib/python3.8/cp/_aes.cpython-38m.so`) from a hard-coded passphrase (in `/service_managesr/service/utils/upgrade_write.pyc`).
+
+It's generated in the following way:
+
+1. The shared lib `_aes.cpython-38m.so` uses [EVP_BytesToKey](https://www.openssl.org/docs/manmaster/man3/EVP_BytesToKey.html) to generate the AES key and IV from a passphrase
+2. The passphrase can be found in ther file `upgrade_write.pyc` (it's `1.57079628`)
+3. Next, the passphrase is used to generate a "pre-key". This can be done with the openssl tool:
+```bash
+$ openssl enc -nosalt -p -in foo -out bar -e -aes256 -k "1.57079628" -md sha1
+```
+4. Next, that "pre-key" is used to decrypt a base64 string. The resulting "plaintext" makes a new passphrase: `LL0QuKHDBrOKeGHUSvUTVpwbHB2WzNcP`
```

Comment:
I would not publish the passphrase but let it as an exercise for the reader ;-)

Reply:
Fixed.